### PR TITLE
Support arbitrary visitors

### PR DESCRIPTION
**Pull request by lisitsyn**

### CODE REVIEW

File: tests/unit/lib/Any_unittest.cc [hunk 1]
```diff
@@ -836,3 +836,31 @@ TEST(Any, cast_into_sgobject)
...
+	};
+	Any::register_visitor<int, ExtractVisitor>([&] (int value, ExtractVisitor* visitor) { extracted_value = value; });
+	auto any = make_any<int>(42);
+	any.visit<ExtractVisitor>();
+	EXPECT_EQ(any.as<int>(), extracted_value);
+}
+
+TEST(Any, stateful_visit)
+{
+	std::string value = "hello, world!";
```
Comment (iglesias) [inline tests/unit/lib/Any_unittest.cc]: const?

File: tests/unit/lib/Any_unittest.cc [hunk 2]
```diff
@@ -836,3 +836,31 @@ TEST(Any, cast_into_sgobject)
...
+	struct StringStreamVisitor
+	{
+		std::stringstream ss;
+	};
+	Any::register_visitor<std::string, StringStreamVisitor>([&] (std::string value, StringStreamVisitor* visitor) {
```
Comment (iglesias) [inline tests/unit/lib/Any_unittest.cc]: const std::string&?

File: tests/unit/lib/Any_unittest.cc [hunk 3]
```diff
@@ -836,3 +836,31 @@ TEST(Any, cast_into_sgobject)
...
+		(visitor->ss) << value;
+	});
+	auto any = make_any(value);
+	StringStreamVisitor visitor;
+	any.visit(&visitor);
```
Comment (iglesias) [inline tests/unit/lib/Any_unittest.cc]: What about a couple of calls to visit to really show the stateful bit?
Perhaps I am misinterpreting the state reference.

File: tests/unit/lib/Any_unittest.cc [hunk 4]
```diff
@@ -836,3 +836,31 @@ TEST(Any, cast_into_sgobject)
...
 	EXPECT_THROW(any.cast<int>(), std::logic_error);
 }
+
+TEST(Any, simple_visit)
+{
+	int extracted_value{};
+	struct ExtractVisitor
+	{
+	};
+	Any::register_visitor<int, ExtractVisitor>([&] (int value, ExtractVisitor* visitor) { extracted_value = value; });
```
Comment (iglesias) [inline tests/unit/lib/Any_unittest.cc]: Just wild thining here. From api pov, I am wondering whether the template arguments could be somehow omitted and extracted from the std::function.

**Review by iglesias — APPROVED** [review]
¡Qué bueno!

I left just mostly nitpicking comments 😅

File: tests/unit/lib/Any_unittest.cc (hunk 4 above)
Comment (lisitsyn) [reply tests/unit/lib/Any_unittest.cc]: Si

File: tests/unit/lib/Any_unittest.cc (hunk 1 above)
Comment (lisitsyn) [reply tests/unit/lib/Any_unittest.cc]: Meh doesn't matter :)

File: tests/unit/lib/Any_unittest.cc (hunk 2 above)
Comment (lisitsyn) [reply tests/unit/lib/Any_unittest.cc]: Replaced by auto

File: tests/unit/lib/Any_unittest.cc (hunk 3 above)
Comment (lisitsyn) [reply tests/unit/lib/Any_unittest.cc]: Well it is already stateful but I can visit twice to make it more visible

File: src/shogun/lib/any.h [hunk 5]
```diff
@@ -1164,26 +1170,67 @@ namespace shogun {
...
+	void Any::register_caster(std::function<To(From)> caster)
+	{
+		const auto key = std::make_pair(std::type_index{typeid(From)}, std::type_index{typeid(To)});
+
+		if (casting_registry.count(key))
+		{
+			return;
+		}
+
+		casting_registry[key] = [caster] (void* src, void* dst)
```
Comment (gf712) [inline src/shogun/lib/any.h:1197]: you already pass the function pointer by value, you might as well std::move here no?

File: src/shogun/lib/any.h [hunk 6]
```diff
@@ -1164,26 +1170,67 @@ namespace shogun {
...
+		{
+			auto typed_src = static_cast<From*>(src);
+			auto typed_dst = static_cast<To*>(dst);
```
Comment (gf712) [inline src/shogun/lib/any.h:1200]: out of curiosity how come you switched from reinterpret_cast to static_cast?

Comment (lisitsyn) [reply src/shogun/lib/any.h:1200]: We don't actually need reinterpret so I used more narrow one.

File: src/shogun/lib/any.h (hunk 5 above)
Comment (lisitsyn) [reply src/shogun/lib/any.h:1197]: You mean caster&&?

**Comment by karlnapf** [comment]
:dancer:

File: src/shogun/lib/any.h (hunk 5 above)
Comment (gf712) [reply src/shogun/lib/any.h:1197]: Yup, [caster=std::move(caster)]

**Pull request merged**